hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataPortableResolver.java

```diff
@@ -37,30 +40,42 @@
 import javax.annotation.Nullable;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.EnumMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.stream.Stream;
 
-import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_CLASS_ID;
-import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_CLASS_VERSION;
-import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_FACTORY_ID;
-import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_CLASS_ID;
-import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_CLASS_VERSION;
-import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_FACTORY_ID;
 import static com.hazelcast.jet.sql.impl.connector.SqlConnector.PORTABLE_FORMAT;
-import static com.hazelcast.jet.sql.impl.connector.SqlConnectorUtil.asInt;
 import static com.hazelcast.jet.sql.impl.connector.keyvalue.KvMetadataResolver.extractFields;
 import static com.hazelcast.jet.sql.impl.connector.keyvalue.KvMetadataResolver.maybeAddDefaultField;
 import static com.hazelcast.sql.impl.extract.QueryPath.KEY;
 import static com.hazelcast.sql.impl.extract.QueryPath.VALUE;
 
-final class MetadataPortableResolver implements KvMetadataResolver {
+public final class MetadataPortableResolver implements KvMetadataResolver {
+    public static final DefaultedMap<FieldType, QueryDataType> PORTABLE_TO_SQL = new DefaultedMap<>(
+            new EnumMap<>(ImmutableMap.<FieldType, QueryDataType>builder()
```

Comment:
> What is the rationale for using `EnumMap` over `switch`? Readability? Performance (I wonder which is better)? Reusability?

I introduced `DefaultedMap` by assuming that Java doesn't perform any optimization for switch-case statements. However, it turned out that compact switches, such as enumerations and (roughly) consecutive integers, are converted into lookup tables; switches of arbitrary numbers are sorted for binary search; and string switches are converted into hash tables. Thus, `DefaultedMap` doesn't bring any performance gain over switches and usually performs worse.

However, for _data mapping_, we can use `DefaultedMap` with `EnumMap`, which is also a lookup table. It will also perform worse than a switch because in switches, the lookup table is hardcoded into the method and lives in the stack space, so that when you call the _resolve_ method, the table is already loaded into the cache. In `EnumMap`, you follow a pointer to access the table and wait for it to be loaded. However, `Map` values are initialized once, which can justify the reduced spatial locality if the values are not "static", as in `case X: return List.of(...)`, by reducing garbage.

In terms of readability, classical switches might win due to their familiarity. However, expressing a mapping using a switch statement seems like an "anti-pattern" to me.

In terms of reusability, `Map`s definitely win since we can transform and combine them to obtain new mappings without repeating the original mapping. In addition, in `DefaultedMap`, you can avoid default value/exception by using `get()` instead of `getOrDefault()`, which enables us to reuse exactly the same mapping for different logic.

Reply:
I think that the last thing (reusability/transfromability/revesibility) is the only important reason for favouring Map over `switch`, especially for enums. I would not say that this is "data mapping" (BTW, there are some improvements for switch like pattern matching, so definitely switch is not obsolete way to do mapping).

We can keep the refactored code in this PR hoping that it will not impact performance much. It should be used only during CREATE MAPPING - is that correct? But generally I would not refactor from switch to Map just for the sake of refactoring.

Reply:
Yes, they are only used in `CREATE MAPPING` statement, and especially not during upserts. `UpsertInjector`s don't do any type lookup after creation.